easybuild/easyblocks/j/jaxlib.py

```diff
@@ -0,0 +1,132 @@
+from easybuild.easyblocks.generic.pythonpackage import PythonPackage
+from easybuild.tools.build_log import EasyBuildError
+from easybuild.tools.filetools import apply_regex_substitutions, which
+from easybuild.tools.modules import get_software_root, get_software_version
+
+
+class EB_jaxlib(PythonPackage):
+    """Support for installing jaxlib. Extension of the existing PythonPackage easyblock"""
+
+    @staticmethod
+    def extra_options():
+        """Custom easyconfig parameters specific to jaxlib."""
+        extra_vars = PythonPackage.extra_options()
+
+        extra_vars['use_pip'][0] = True
+        # Run custom build script and install the generated whl file
+        extra_vars['buildcmd'][0] = '%(python)s build/build.py'
+        extra_vars['install_src'][0] = 'dist/*.whl'
+
+        return extra_vars
+
+    def configure_step(self):
+        """Custom configure step for jaxlib."""
+
+        super(EB_jaxlib, self).configure_step()
+
+        binutils_root = get_software_root('binutils')
+        if not binutils_root:
+            raise EasyBuildError("Failed to determine installation prefix for binutils")
+        config_env_vars = {
+            # This is the binutils bin folder: https://github.com/tensorflow/tensorflow/issues/39263
+            'GCC_HOST_COMPILER_PREFIX': os.path.join(binutils_root, 'bin'),
+        }
+
+        # Collect options for the build script
+        # Used only by the build script
+        options = [
+            '--target_cpu_features=default',  # Using copt for optimizations
+        ]
+        # Passed directly to bazel
+        bazel_startup_options = [
+            '--output_user_root=%s' % tempfile.mkdtemp(suffix='-bazel', dir=self.builddir),
+        ]
+        # Passed to the build command of bazel
+        bazel_options = [
+            '--jobs=%s' % self.cfg['parallel'],
+            '--subcommands',
+            '--action_env=PYTHONPATH',
+            '--action_env=EBPYTHONPREFIXES',
+        ]
+        if self.toolchain.options.get('debug', None):
+            bazel_options.extend([
+                '--strip=never',
+                '--copt="-Og"'
+            ])
+        # Add optimization flags set by EasyBuild each as a separate option
+        bazel_options.extend(['--copt=%s' % i for i in os.environ['CXXFLAGS'].split(' ')])
+
+        cuda_root = get_software_root('CUDA')
+        if cuda_root:
+            cudnn_root = get_software_root('cuDNN')
+            if not cudnn_root:
+                raise EasyBuildError('For CUDA-enabled builds cuDNN is also required')
+            cuda_version = '.'.join(get_software_version('CUDA').split('.')[:2])  # maj.minor
+            cudnn_version = '.'.join(get_software_version('cuDNN').split('.')[:3])  # maj.minor.patch
+            options.extend([
+                '--enable_cuda',
+                '--cuda_path=' + cuda_root,
+                '--cuda_compute_capabilities=' + self.cfg.template_values['cuda_compute_capabilities'],
```

Comment:
Rather than accessing `self.cfg.template_values['cuda_compute_capabilities']` directly, we should probably provide a wrapper function in framework to grab the correct value for `cuda_compute_capabilities` based on the `--cuda-compute-capabilities` EasyBuild configuration option and the `cuda_compute_capabilities` easyconfig parameter.

The problem with this approach is that you get a hard crash when `cuda_compute_capabilities` is not set in either way, with:

```
KeyError: 'cuda_compute_capabilities'
```

Reply:
Accessing the template values has the benefit that you can take whatever format you need, i.e. space-separated, comma-separated, sm-style etc. and all that already correctly and consistently handled (i.e. EC vs cfg option)
I mean: We already have that wrapper function: It sets `self.cfg.template_values`: https://github.com/easybuilders/easybuild-framework/blob/ef658b78b7d1a26da7c685c5e3b5f2afa9d3057d/easybuild/framework/easyconfig/templates.py#L351
The only benefit of a function would be a slightly better error in case this is unset, but given the different possible formats I don't think this is worth the additional code.

We should probably set the value in the ECs as default values to avoid this being unset. And then IMO `KeyError: 'cuda_compute_capabilities'` is a good enough error for a faulty EC to point to the issue.

Reply:
The problem is that is isn't just the `KeyError`, there's a whole traceback that goes along with it.

Letting EasyBuild crash, and not mentioning where/how the problem can be fixed, isn't exactly good error reporting.

So we should first check whether the value is there (`if 'cuda_compute_capabilities' in self.cfg.template_values`), and if not produce a clear/clean error message.

And since this is a generic thing, not specific to `jaxlib` at all, this could be done by a wrapper function in framework, which can then also be used in other easyblocks like TensorFlow.

Reply:
Depends on who the error reporting is for. I find those tracebacks and errors from Python quite helpful, at least in such cases.
And the error is impossible if the ECs have cuda_compute_capabilities in them which are used by default. So the error would be for us testers or the devs of ECs.
But up to you.

Unfortunately I currently don't have time to implement that function myself, so maybe leave as-is and improve on this in a follow-up?

Reply:
Implementing something that can be easily leveraged from an easyblock is fairly easy, done in https://github.com/easybuilders/easybuild-framework/pull/3807

Reply:
@Flamefire https://github.com/easybuilders/easybuild-framework/pull/3807 is merged, so we can change this to:

```suggestion
                '--cuda_compute_capabilities=' + self.cfg.get_cuda_cc_template_value('cuda_compute_capabilities'),
```